File: src/main/scala/lang/CapiQueryString.scala
package cql.lang

import TokenType.*
import cql.lang.*

import java.net.{URI, URLEncoder}
import java.nio.charset.Charset
import scala.collection.mutable.Map
import scala.collection.mutable.ListBuffer

case class CapiCqlStringError(message: String) extends Error(message)

object CapiCqlString {
  def build(
      program: QueryList
  ): String =
    val (searchStrs, otherQueries) = program.exprs.partitionMap {
      case q: CqlBinary => Left(strFromBinary(q))
      case CqlField(key, Some(value)) =>
        Right(s"${key.literal.getOrElse("")}=${value.literal.getOrElse("")}")
      case CqlField(key, None) =>
        throw new CapiCqlStringError(
          s"The field '+$key' needs a value after it (e.g. +$key:tone/news)"
        )
      case QueryOutputModifier(key, Some(value)) =>
        Right(s"${key.literal.getOrElse("")}=${value.literal.getOrElse("")}")
      case QueryOutputModifier(key, None) =>
        throw new CapiCqlStringError(
          s"The output modifier '@$key' needs a value after it (e.g. +$key:all)"
        )
    }

    val maybeSearchStr = searchStrs match {
      case Nil => None
      case strs =>
        val encodedString =
          new URI(null, null, strs.mkString(" "), null).toASCIIString
        Some(s"q=$encodedString")
    }

    List(maybeSearchStr, otherQueries).flatten.mkString("&")

  private def strFromContent(queryContent: QueryExpr): String =
    queryContent.content match {
      case CqlStr(str)       => str
      case CqlGroup(content) => s"(${strFromBinary(content)})"
      case q: CqlBinary      => strFromBinary(q)
    }

  private def strFromBinary(queryBinary: CqlBinary): String =
    val leftStr = strFromContent(queryBinary.left)
    val rightStr = queryBinary.right
      .map { case (op, content) =>
        s" ${op.tokenType.toString} ${strFromBinary(content)}"
      }
      .getOrElse("")
    leftStr + rightStr
}

File: client/src/lang/capiQueryString.ts
import { err, ok, Result } from "../utils/result";
import { CqlQuery, CqlBinary, CqlExpr } from "./ast";
import { getCqlFieldsFromCqlBinary } from "./utils";

class CapiCqlStringError extends Error {
  public constructor(message: string) {
    super(message);
  }
}

const dateFields = ["from-date", "to-date"];
const relativeDateRegex = /-(?<quantity>\d+)(?<unit>[dmyw])/;

const parseDateValue = (value: string): string => {
  const result = relativeDateRegex.exec(value);
  if (!result) {
    return value;
  }
  const now = new Date();
  const { quantity, unit } = result.groups as {
    quantity: string;
    unit: string;
  };

  const year = now.getFullYear() - (unit === "y" ? parseInt(quantity) : 0);
  // Months are zero indexed in Javascript, ha ha ha
  const month = now.getMonth() - (unit === "m" ? parseInt(quantity) : 0);
  const day =
    now.getDate() -
    (unit === "d"
      ? parseInt(quantity)
      : unit === "w"
        ? parseInt(quantity) * 7
        : 0);
  const date = new Date(year, month, day);
  return `${date.getFullYear()}-${(date.getMonth() + 1).toString().padStart(2, "0")}-${date.getDate().toString().padStart(2, "0")}`;
};

export const queryStrFromQueryList = (
  query: CqlQuery
): Result<Error, string> => {
  const { content } = query;
  if (!content) {
    return ok("");
  }

  const searchStrs = strFromBinary(content);

  try {
    const otherQueries = getCqlFieldsFromCqlBinary(content).flatMap((expr) => {
      switch (expr.type) {
        case "CqlField": {
          if (expr.value) {
            const value = dateFields.includes(expr.key.literal ?? "")
              ? parseDateValue(expr.value?.literal ?? "")
              : expr.value.literal;
            return [`${expr.key.literal ?? ""}=${value}`];
          } else {
            throw new CapiCqlStringError(
              `The field '${expr.key.literal}' needs a value after it (e.g. '${expr.key.literal}:tone/news')`
            );
          }
        }
        default: {
          return [];
        }
      }
    });

    const queryStr = searchStrs.length
      ? `q=${encodeURI(searchStrs.trim())}`
      : "";

    return ok([queryStr, otherQueries].filter(Boolean).flat().join("&"));
  } catch (e) {
    return err(e as Error);
  }
};

const strFromContent = (queryContent: CqlExpr): string | undefined => {
  const { content } = queryContent;
  switch (content.type) {
    case "CqlStr":
      return content.searchExpr;
    case "CqlGroup":
      return `(${strFromBinary(content.content).trim()})`;
    case "CqlBinary":
      return strFromBinary(content);
    default:
      // Ignore fields
      return;
  }
};

const strFromBinary = (queryBinary: CqlBinary): string => {
  const leftStr = strFromContent(queryBinary.left);

  const rightStr =
    queryBinary.right &&
    // Something of a hack — don't include
    queryBinary.right.binary.left.content.type !== "CqlField"
      ? `${queryBinary.right.operator} ${strFromBinary(queryBinary.right.binary)}`
      : "";

  return (leftStr ?? "") + (rightStr ? ` ${rightStr.trim()} ` : "");
};
